proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,649 @@
+work on Dependent Haskell.
+
+To summarize, there are three reasons to make this change:
+
+* Language consistency (symmetry between terms and types)
+* Ability to design better APIs (good error messages, no proxy types, no ambiguous types)
+* Prepare the compiler internals for further work on dependent types
+
+Proposed Change Specification
+-----------------------------
+
+1. Add a new language extension, ``RequiredTypeArguments``.
+
+2. When ``RequiredTypeArguments`` is in effect, lift the restriction that the
+   ``forall a ->`` quantifier cannot be used in types of terms.
+
+3. In types of terms, ``forall a ->`` is an irrelevant quantifier.
+
+4. Extend the term-level syntax with several constructs that previously could
+   only occur at the type level:
+
+   * Function arrows: ``a -> b``
+   * Multiplicity-polymorphic function arrows: ``a %m -> b`` (under ``-XLinearTypes``)
+   * Constraint arrows: ``a => b``
+   * Universal quantification: ``forall a. b``
+   * Visible universal quantification: ``forall a -> b``.
+
+   We will call them "types-in-terms". They are subject to the following rules:
+
+   * Grammatically, their constituents are terms,
+     not types::
+
+                proposed grammar:                 as opposed to:
+         +-------------------------------+----------------------------------+
+         |                               |                                  |
+         |  exp -> ...                   |    exp -> ...                    |
+         |       | exp -> exp            |         | type -> type           |
+         |       | exp => exp            |         | type => type           |
+         |       | forall bndrs. exp     |         | forall bndrs. type     |
+         |       | forall bndrs -> exp   |         | forall bndrs -> type   |
+         |                               |                                  |
+         +-------------------------------+----------------------------------+
+
+     This is a necessity to avoid parsing conflicts. As a consequence, the
+     ``'`` symbol signifies Template Haskell name quotation rather than
+     ``DataKinds`` promotion, and ``*`` is treated as an infix operator
+     regardless of ``-XStarIsType``.
+
+     For the above grammar to work, make ``forall`` a keyword at the term
+     level. This is to be guarded by ``RequiredTypeArguments``, but only for
+     three releases (see point 9 below).
+
+   * In name resolution, we follow the term-level rules that prioritize data
+     constructors over type constructors, and variables over type variables.
+
+     The advantage is that we have less subtle rules this way, so that in ``f
+     T`` and in ``f (T -> T)``, the ``T`` resolves to the same identifier.
+
+     The disadvantage is that in ``\x -> f (forall x. x)``, the ``x``
+     occurrence refers to the lambda-bound variable rather than the
+     ``forall``-bound type variable, even though the ``forall`` binding is
+     closer to the occurrence.
```

Comment:
There are several alternatives we could take here, which I outlined in https://github.com/ghc-proposals/ghc-proposals/pull/281#issuecomment-702981186

Regarding your suggestion of an error, this would make the extension "fork-like" (i.e. enabling it makes GHC reject currently valid programs). Instead, we could add a warning, which is the subject of a separate proposal, #270.

Reply:
I'm not sure a warning is quite good enough, since you would get errors before the warning shows up.

At any rate, I was suggesting that this very situation which couldn't happen before. Therefore, it shouldn't break compatibility, but maybe it's too hairy to specify.

By the way, the alternatives you cite should be covered in the alternatives section.

Reply:
> At any rate, I was suggesting that this very situation which couldn't happen before. Therefore, it shouldn't break compatibility, but maybe it's too hairy to specify.

Maybe I'm misunderstanding what exactly you suggest to turn into an error, then, because it's already possible to have a type variable and a term variable that share the same name.

> By the way, the alternatives you cite should be covered in the alternatives section.

Yes, as soon as the committee tells me which options go into "Proposed Change Specification", and which ones go into "Alternatives". I'm not looking forward to rewriting the text several times.